ktor-swagger/src/main/kotlin/de/nielsfalk/playground/ktor/swagger/Swagger.kt

```diff
@@ -219,24 +232,136 @@ val propertyTypes = mapOf(
     Date::class to Property("string", "date-time"),
     LocalDateTime::class to Property("string", "date-time"),
     LocalDate::class to Property("string", "date")
-).mapKeys { it.key.qualifiedName }.mapValues { it.value to emptyList<KClass<*>>() }
-
-fun <T, R> KProperty1<T, R>.toModelProperty(): Pair<Property, Collection<KClass<*>>> =
-    (returnType.classifier as KClass<*>)
-        .toModelProperty(returnType)
-
-private fun KClass<*>.toModelProperty(returnType: KType? = null): Pair<Property, Collection<KClass<*>>> =
-    propertyTypes[qualifiedName?.removeSuffix("?")]
-        ?: if (returnType != null && toString() == "class kotlin.collections.List") {
-            val kClass: KClass<*> = returnType.arguments.first().type?.classifier as KClass<*>
-            val items = kClass.toModelProperty()
-            Property(items = items.first, type = "array") to items.second
+).mapKeys { it.key.qualifiedName }.mapValues { it.value to emptyList<TypeInfo>() }
+
+fun <T, R> KProperty1<T, R>.toModelProperty(reifiedType: Type? = null): Pair<Property, Collection<TypeInfo>> =
+    returnType.toModelProperty(reifiedType)
+
+fun KType.toModelProperty(reifiedType: Type?): Pair<Property, Collection<TypeInfo>> =
+    resolveTypeInfo(reifiedType).let { typeInfo ->
+        val type = typeInfo?.type ?: classifier as KClass<*>
+        type.toModelProperty(this, typeInfo?.reifiedType as? ParameterizedType)
+    }
+
+internal fun <T, R> KProperty1<T, R>.returnTypeInfo(reifiedType: Type?): TypeInfo =
+    returnType.resolveTypeInfo(reifiedType)!!
+
+internal fun KType.resolveTypeInfo(reifiedType: Type?): TypeInfo? {
```

Comment:
Does this need to be nullable? It looks like the only nullable case is `KClass<*>` which is then immediately non-null asserted.

Reply:
I answered offline, but for documentation purposes, yes. Null is used.